### ISSUE TITLE
alligator point, fl

### ISSUE THREAD

**Issue by b-may** [pelias/pelias#464]
Hey team! 

I was using your awesome geocoding engine when I noticed something interesting.
Let me tell you more about it.

---
#### Here's what I did :innocent:

http://pelias.github.io/compare/#/v1/search%3Ftext=alligator%20point,fl

---
#### Here's what I got :scream_cat:

http://pelias.github.io/compare/#/v1/search%3Ftext=alligator%20point,fl

---
#### Here's what I was expecting :sparkles:

Find alligator point, fl

---
#### Here's what I think could be improved :trophy:
Tried `bald point, fl` too in same area, same problem. Use GNIS lookups first before sending to libpostal?

**Comment by b-may** [comment]
I'm finding a fair number of libpostal address parsing problems (I am assuming so based on the query object). 

For example, here's another one:

http://pelias.github.io/compare/#/v1/search%3Ftext=103%20BEAL%20PKWY%20SE,%20FT%20WALTON%20BEACH,%20FL

<pre>"query": {
      "text": "103 BEAL PKWY SE, FT WALTON BEACH, FL",
      "parsed_text": {
        "number": "103",
        "street": "ft",
        "query": "se",
        "neighbourhood": "walton",
        "city": "beach",
        "state": "fl"
      },
</pre>

Should I keep opening issues with more examples? Just stuff more examples into one issue? Or should I open issues over at libpostal?

**Comment by trescube** [comment]
The libpostal team is actively working on a new parsing model which should address a number of issues that we're seeing with parsing.  Since a number of the issues you're seeing are limited to libpostal, opening issues at https://github.com/openvenues/libpostal would remove us as a middle layer and you'd probably get feedback a lot faster on when you can expect these issues to be fixed.  We keep an active eye on libpostal issues, too.

**Comment by b-may** [comment]
OK, thanks. 

I'm reading more about how pelias works and ran across your libpostal api and gave it a shot:

<pre>
curl -s 'https://libpostal.mapzen.com/parse?address=103+BEAL+PKWY+SE,+FT+WALTON+BEACH,+FL' | python -mjson.tool
[
    {
        "label": "house_number",
        "value": "103"
    },
    {
        "label": "road",
        "value": "beal pkwy se"
    },
    {
        "label": "city",
        "value": "ft walton beach"
    },
    {
        "label": "state",
        "value": "fl"
    }
]
</pre>

So why does that work, but the output from search does not?

**Comment by trescube** [comment]
That's a very good question, I'm not entirely sure but I'll look into it.

**Comment by orangejulius** [comment]
It looks like this is a regression in the pre-release build of libpostal we use in Mapzen Search. I started up an Ubuntu VM and installed the default build of libpostal, and got this:

```
> 103 BEAL PKWY SE, FT WALTON BEACH, FL

Result:

{
  "house_number": "103",
  "road": "beal pkwy se",
  "city": "ft walton beach",
  "state": "fl"
}
```

This one is definitely worth reporting to libpostal to ensure it works again in the future.

**Comment by trescube** [comment]
Yeah, I agree with @orangejulius, we use a different build of libpostal than what's hosted.  In general, what we use parses more accurately but unfortunately it seems that there may be some regressions.

**Comment by thatdatabaseguy** [comment]
Checked and it's more of a data sparsity issue. That's a fairly rare city in terms of the number of OSM addresses contained therein.

Essentially the parser generates and memorizes a large dictionary of the city names it sees in OSM addresses during training, and that's one of the strongest signals used to make a prediction for a given word/phrase. Obviously dictionary memorization is not sufficient on its own e.g. not every instance of the phrase "Brooklyn" in fact refers to the borough - can be "Brooklyn Avenue" or "Brooklyn Museum", etc. - but that's fine because surrounding words/phrases are used in the prediction as well.

Because OSM uses full unabbreviated names for everything (see "[Fort Walton Beach](http://www.openstreetmap.org/relation/118885)"), and real addresses do not, the strategy is to abbreviate toponyms some fraction of the time, about 30% of the time. That's not a problem for something common like "Fort Lauderdale" vs. "Ft Lauderdale", but can be for very rare city names. Since only 2 addresses matched the polygon for Ft Walton Beach in the training data, it happened that by random chance the phrase "Ft Walton Beach" occurred once in the original libpostal training data and zero times when the OSM data was ingested again to build the intermediate model.

The solution to things like this is in the works for the next version of the model, which is trained not only on addresses, but on simple city names and their parent admins as well (including cities in less-well-mapped places that are stored as points rather than polygons in OSM). The places-only data set guarantees that every place in the world will be listed in a minimum of 5 training examples with both abbreviated and unabbreviated forms ("Mount Vernon" and "Mt Vernon", etc.), plus some extra examples proportional to its population.

FYI, one trick that can be used in the mean time to test whether a given place name made it into libpostal's dictionaries is to type the name standalone into the parser's command-line and see what it returns. If it returns the entire input with a single label like {suburb, city_district, city, state_district, state, country}, then the parser has almost definitely seen that place name before. There's a heuristic in the parser where, if the input is a single known phrase, it will forgo using the sequence model altogether and simply return the most likely label for that phrase as estimated from the input data (so if the phrase "New York" was more commonly labeled a state than a city, it will return "state"). This can also help identify blind spots i.e. places/name variations that are absent from the input data.

**Comment by b-may** [comment]
@thatdatabaseguy - thanks for the detailed explanation!

**Comment by thatdatabaseguy** [comment]
No worries. Also, two more updates:

1. [OSM polygons in Florida are insane](http://www.openstreetmap.org/relation/117942) (city-level gerrymandering or something?). Have made a few improvements to the libpostal OSM polygon index. Sometimes fixing an invalid polygon in GEOS with buffer(0) works as expected and sometimes it ends up creating a valid but useless multipolygon, whereas the original would suffice for point-in-polygon tests. Libpostal's polygon index now automatically tests which version would be better and keeps that one, so more of the "exotic" city polygons should make it in to the address training set when reverse-geocoding.
2. Abbreviation of toponyms was happening less frequently than it was supposed to (random sampling was being done twice so was only happening ~10% of the time). For the next release that will be fixed, which should affect this issue.

**Comment by b-may** [comment]
Interesting. I can help with fixing city boundaries in OSM in Florida if that would help. I've done some work on this in the past, but didn't have a good reason to put tons of effort into it. 

@trescube , @orangejulius - does Pelias rely on OSM admin polygons too?

**Comment by orangejulius** [comment]
Pelias doesn't currently use OSM polygons, but of course anything that helps improve them is a good thing!

**Comment by dianashk** [comment]
Libpostal 1.0 fixes this!

![screen shot 2017-04-20 at 3 40 16 pm](https://cloud.githubusercontent.com/assets/5446511/25249384/b686bad8-25df-11e7-87ed-4055e5bc7011.png)

![screen shot 2017-04-20 at 3 41 30 pm](https://cloud.githubusercontent.com/assets/5446511/25249453/e31cf9ae-25df-11e7-8879-e011e30fd7b2.png)

Close issue after libpostal 1.0 is in production. 🎉

**Comment by b-may** [comment]
Awesome!
